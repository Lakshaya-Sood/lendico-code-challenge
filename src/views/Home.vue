<template>
  <div class="home">
    <b-card class="widget">
      <loan-calculator-widget
        :yearRange="this.yearRange"
        :amountRange="this.amountRange"
        :yearsValue="this.years"
        :amountValue="this.amount"
        @onYearChange="this.handleYearChange"
        @onAmountChange="this.handleAmountChange"
        @onSubmit="this.calculateMonthlyInstallment"
      />
      <br />
      <show-result
        :description="'Monthly installment:'"
        :failMessage="'Compute Again!!!'"
        :result="this.monthlyInstallment"
        :isLoading="this.isCalculating"
      />
    </b-card>
  </div>
</template>

<script lang="ts">
import Vue from "vue";
import { Prop, Component, Emit } from "vue-property-decorator";
import { mapActions, mapGetters } from "vuex";

import LoanCalculatorWidget from "../components/LoanCalculatorWidget.vue";
import ShowResult from "../components/ShowResult.vue";

import { __YearRange__, __AmountRange__ } from "../constants/HomeConstants";

@Component({
  name: "home",
  components: {
    LoanCalculatorWidget,
    ShowResult
  },
  computed: {
    ...mapGetters([
      "years",
      "amount",
      "monthlyInstallment",
      "isCalculating",
      "error"
    ])
  },
  methods: {
    ...mapActions(["onYearChange", "onAmountChange"])
  }
})
export default class Home extends Vue {
  get yearRange() {
    return __YearRange__;
  }
  get amountRange() {
    return __AmountRange__;
  }

  @Emit("onSubmit")
  async calculateMonthlyInstallment() {
    const isSuccess = await this.$store.dispatch("calculateMonthlyInstallment");
    isSuccess
      ? this.$bvToast.toast("Calculated Successfully", {
          title: "Calculate Monthly Installment",
          variant: "success",
          autoHideDelay: 1500,
          solid: true
        })
      : this.$bvToast.toast("Server Error", {
          title: "Calculate Monthly Installment",
          variant: "danger",
          autoHideDelay: 1500,
          solid: true
        });
  }

  @Emit("onYearChange")
  handleYearChange(value: number): void {
    this.$store.dispatch("onYearChange", value);
  }

  @Emit("onAmountChange")
  handleAmountChange(value: number) {
    this.$store.dispatch("onAmountChange", value);
  }
}
</script>

