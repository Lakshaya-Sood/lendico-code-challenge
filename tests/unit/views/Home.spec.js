import { shallowMount, createLocalVue } from '@vue/test-utils'
import BootstrapVue from 'bootstrap-vue'
import Vuex from "vuex";
import Home from '@/views/Home.vue'

const vue_mock = createLocalVue();
vue_mock.use(BootstrapVue);
vue_mock.use(Vuex);
let actions__mock = {
    calculateMonthlyInstallment: jest.fn(),
    onYearChange: jest.fn(),
    onAmountChange: jest.fn()
};
let actions_spy = {
    one: jest.spyOn(actions__mock, 'calculateMonthlyInstallment'),
    two: jest.spyOn(actions__mock, 'onYearChange'),
    three: jest.spyOn(actions__mock, 'onAmountChange'),
}
let getters__mock = {
    years: () => 1,
    amount: () => 100,
    monthlyInstallment: () => null,
    isCalculating: () => false,
    error: () => null
};
let store__mock = new Vuex.Store({
    getters: getters__mock,
    actions: actions__mock
});

describe('home component', () => {
    const wrapper = shallowMount(Home, {
        localVue: vue_mock,
        store: store__mock
    })
    test('name in kebab case', () => {
        expect(wrapper.vm.$options.name).toEqual('home')
    })

    test('[method] calculateMonthlyInstallment', () => {
        wrapper.vm.calculateMonthlyInstallment()
        expect(actions_spy.one).toHaveBeenCalled();
    })

    test('[method] handleYearChange', () => {
        wrapper.vm.handleYearChange()
        expect(actions_spy.two).toHaveBeenCalled();
    })

    test('[method] handleAmountChange', () => {
        wrapper.vm.handleAmountChange()
        expect(actions_spy.three).toHaveBeenCalled();
    })

    test('no changes as compared to last snapshot', () => {
        expect(wrapper.element).toMatchSnapshot() // https://vue-test-utils.vuejs.org/guides/#snapshot-testing
    })
})
