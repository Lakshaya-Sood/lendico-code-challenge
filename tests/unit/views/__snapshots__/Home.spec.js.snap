// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`home component no changes as compared to last snapshot 1`] = `
<div
  class="home"
>
  <b-card-stub
    bodytag="div"
    class="widget"
    footertag="div"
    headertag="div"
    subtitle=""
    subtitletag="h6"
    subtitletextvariant="muted"
    tag="div"
    title=""
    titletag="h4"
  >
    <loan-calculator-widget-stub
      amountrange="[object Object]"
      amountvalue="100"
      yearrange="[object Object]"
      yearsvalue="1"
    />
     
    <br />
     
    <show-result-stub
      description="Monthly installment:"
      failmessage="Compute Again!!!"
    />
  </b-card-stub>
</div>
`;
